### ISSUE TITLE
race condition in AspireUpdate\Debug::log 

### ISSUE THREAD

**Issue by chuckadams** [aspirepress/aspireupdate#187]
Roughly here:
https://github.com/aspirepress/AspireUpdate/blob/e855143dea6e744d6d2c04ff9c55031971c46670/includes/class-debug.php#L154

Reading the whole log, appending a line, and writing the whole thing back is not only extremely inefficient, it has an inherent race condition between the read and the write.  File I/O being what it is, the window of that race condition may be a lot bigger than you think.  The WP filesystem API doesn't have any way to open a file for appending, so it's probably to best just not use that API for log files.

**Comment by namithj** [comment]
Looks like I we are going to have to extend WP-Filesystem class.  We can also submit to core but they probably won't accept since this is an obvious use case which haven't yet been added in.

**Comment by costdev** [comment]
The Filesystem API didn't have much in the way of maintainers over the years. I took over its maintenance a while back, but resolving issues took precedence over new features, and resolving issues in such an old, mission-critical API _is a doozy_ with the BC promise and X number of server configurations.

If we're not planning to support FTP-based or SSH-based filesystems, consider just using native PHP functions.

**Comment by namithj** [comment]
I guess We will have to do both fopen and file_put_contents to cover most of our use cases.  What about inheriting the Filesystem API class and adding functions for these two paths.
Or ignore the class and do it all custom?

**Comment by costdev** [comment]
For now, I'd say we should probably go with the native PHP functions rather than extending the API, as this will need investigation and testing for supporting FTP/SSH systems.

**Comment by chuckadams** [comment]
I'm not sure maintaining a local debug log over ftp/ssh is a great idea anyway.  I mean there's stuff like sshfs we can't do anything about, but filesystems like that are what exacerbate race conditions in the first place.

**Comment by afragen** [comment]
Is there some reason we need a separate `debug.log`? If it's to designate AU errors we could prefix the errors like `[AspireUpdate]`.

**Comment by costdev** [comment]
It would mean loading a full `debug.log` and parsing for `[AspireUpdate]`, which would load a larger file into memory.

Additional note: If an installation has `wp-content` locked down for changes, then this could run into an issue. Usually, plugins would store created files into `wp-content/uploads`, which is usually kept open.

**Comment by namithj** [comment]
What's the opinion about logging in the DB?

**Comment by chuckadams** [comment]
I like the idea of logging in the db, but it seems to be kind of foreign to the WP ecosystem, might not be palatable to some.  error_log sucks() as a logging API, but it seems WP already gets by using only that.  what do the big plugins like yoast or woo do logging?

Might want to distinguish between "useful to know" debug logging from the official record of actions taken: I could get behind creating a new db table for the latter and using that, but not throwing every bit of logging chatter into it.

**Comment by namithj** [comment]
We can throw our log into a non public custom post type.  It won't be amount my first 10 choices but it's still better than going direct file access way.  error_log doesn't have a non hacky way to specify the error log file during each write, else it would have done the job.

We have limited to view log to 1000 items, may be we only save the latest 1000 items.

**Comment by costdev** [comment]
> What do the big plugins like yoast or woo do for logging?

WooCommerce has two options:
![image](https://github.com/user-attachments/assets/c9f40c14-baaf-41e9-8df7-dce106d460d9)

**Filesystem** stores to `wp-content/uploads/wc-logs/`

Not sure what if anything Yoast does

**Comment by namithj** [comment]
We should check what approach woocommerce is using to write to the file log.

**Comment by costdev** [comment]
- Reading: [Uses fopen()](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/Admin/Logging/FileV2/File.php#L267) and [outputs here](https://github.com/woocommerce/woocommerce/blob/184ccb2f8a3edb2d167c5142ecdffeca7d3d4db3/plugins/woocommerce/src/Internal/Admin/Logging/PageController.php#L392-L400)
- Writing: [Uses fopen( '', 'ab' ) and fwrite()](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/Admin/Logging/FileV2/File.php#L432)

**Comment by costdev** [comment]
Yeah, and also a reminder of the location Woo uses: `wp-content/uploads/wc-logs/`. We should do the same for AspireUpdate.

**Comment by namithj** [comment]
Do you have bandwidth to take this over?

**Comment by costdev** [comment]
Not at the moment. I'm away all weekend and not sure what bandwidth I have during next week yet.

**Comment by costdev** [comment]
@namithj @chuckadams Aside from the location of the log file, has the primary issue now been resolved by #215?

**Comment by chuckadams** [comment]
@costdev LGTM.  I haven't had the time to give it a lot of testing, but from what I can see the new FS class can open for append, that's what the AU logging is doing, and that should be good.

**Comment by namithj** [comment]
Closed via #215 and #201
